### PR TITLE
Fix #5835 Durable filters and servlets

### LINKED ISSUE #5835: Review Durable Filters, Servlets and Listeners
**Jetty version**

9.4.x

**Description**

Several issues have been seen with durable filters, servlets and listeners:

- Filters, Servlets and Listeners added via the embedded API are assumed to be durable, however if they are added during starting phase, they should be assumed to be non durable
- non durable ServletContextListeners are removed before the contextDestroyed event is called, so it is never called for those listeners

### PR DESCRIPTION
**Pull request by gregw**
Fix #5835 Durable filters and servlets with a general ServletHandler cleanup.
This replaces #5837

### CODE REVIEW

**Comment by gregw** [comment]
@janbartel nudge

**Comment by gregw** [comment]
@janbartel re-nudge

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java [hunk 1]
```diff
@@ -226,113 +242,93 @@ protected void stop(LifeCycle l) throws Exception
...
-            FilterMapping[] fms = filterMappings.toArray(new FilterMapping[0]);
-            updateBeans(_filterMappings, fms);
-            _filterMappings = fms;
+            //Retain only durable filters and mappings
+            updateBeans(_filters, filterHolders);
+            _filters.clear();
+            _filters.addAll(filterHolders);
 
-            _matchAfterIndex = (_filterMappings.length == 0 ? -1 : _filterMappings.length - 1);
+            // Update indexes for prepending filters
```
Comment (janbartel) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java]: I don't see any code here that removes the non-durable filtermappings, yet we've set the matchBefore and matchAfter index.

**Review by janbartel — CHANGES_REQUESTED** [review]

**Review by janbartel — APPROVED** [review]
If CI is happy, then ship it!

**Pull request merged**